### ISSUE TITLE
Libretro core

### ISSUE THREAD

**Issue by Margen67** [joncampbell123/dosbox-x#893]
I noticed there's a RetroArch core for the original DOSBox, but not DOSBox-X.
I found a request on [reddit](https://www.reddit.com/r/RetroArch/comments/80nn2z/suggestion_dosboxx_core/) but nothing on the issue tracker.
Much like the creator of the reddit post, I would also like to see DOSBox-X core as it's better than the original DOSBox.

**Comment by joncampbell123** [comment]
Sounds good to me, but I personally don't have any interest in implementing it at this time.

What does it take to add DOSBox-X to the libretro project?

**Comment by Margen67** [comment]
I'm not sure, but perhaps these might help:
https://github.com/libretro/dosbox-libretro
https://www.libretro.com/index.php/develop/

**Comment by sikthehedgehog** [comment]
Probably a good start would be to make a list of what had been stripped in that DOSBox fork (I imagine that it's just about everything UI, since that'd be RetroArch's job?). Even if nothing comes out of it, at least it could give a good clue as to what could use with more decoupling if needed (to help untie the core from the interface).

**Comment by fr500** [comment]
no the stripped core strips SDL basically because they don't like dependencies.
It wasn't a good reason and in the long run it hurt the core (no auto cycles, no cdrom, stubbed timers, stubbed threads)

The SDL mapper can go, the rest can be kept, no reason to remove things.

As far as I can see the general architecture seems to be mostly the same, so the same skeleton could be used (from the SVN core)

**Comment by yksoft1** [comment]
@fr500 It has dynamic core but only for x86-32, not for x86-64 or arm platforms.

**Comment by fr500** [comment]
Ah, so dynamic_x86, not dynrec
I thought the new dynrec was less hacky, more accurate, even though it's slower.

DYNREC is by far slower but it's quite nice in that it can accomodate a bunch of different backends.
I have my core running with dynrec in AARCH64

I'll take a stab at making a simple libretro port this weekend if I can figure out building under msys2

**Comment by fr500** [comment]
Seems a number of internal things have changed, so the architecture is not.. identical
I tried a port but ran into issues, control->Init(), etc stuff is different now. mixer has some new members in some functions, etc.

Furthermore the current dosbox libretro is a gigantic hack on top of the surface renderer.
Does dosbox iterate per-frame? (my guess is not)

I think a completely new port would be better than trying to

**Comment by evbo** [comment]
@fr500 have you looked at dosbox-svn libretro core? It is the state of the art for libretro from what I understand and might offer more insight:

https://git.libretro.com/libretro/dosbox-svn

Having dosbox-x as a libretro core would be amazing. You could run win95 on [Lakka](http://www.lakka.tv/)! I already do this with win3.1 and it works great.

**Comment by Anthony3yygfvfbh** [comment]
Dosbox-x on retroarch would be great

**Comment by coppermine733** [comment]
I would like to see this as well

**Comment by Ric-82** [comment]
I'd like to see dosbox-x as retroarch core too, especially if it will run on Android too.

**Comment by RaouleMenard** [comment]
I'm also looking forward for Dosbox-X as a RetroArch core.

For Android support with gamepad - as I plan to use it on a Retroid Pocket 2+.

Mostly for playing Windows 95/98 games and Nec PC98 support - RetroArch already have the Dosbox Pure core which is perfect for normal Dos, but not for Windows nor PC98.

**Comment by evbo** [comment]
Here's another reason dosbox x would be well received by the libretro community:
https://github.com/schellingb/dosbox-pure/issues/105

In that issue you can see the trials and tribulations of poor souls trying to get dosbox pure or svn (both have cores) trying to run win9x games.

There'd be plenty of fanfare :)

**Comment by evbo** [comment]
This is not a panacea by any means, but if you're willing to let go of some conveniences libretro cores provide, I just found this core that will run arbitrary bash commands:
https://github.com/SwedishGojira/libretro-bash-launcher

What this means is you can run it just like any other core and then start `dosbox-x` as a standalone Linux process rather than libretro core. Not perfect, but dang near close!

**Comment by Kreijstal** [comment]
ahhh yes, this would be nice! I have some questions, libretro is like a framework right? The problem is mainly writing the interface between libretro and dosbox-x to create a core, am I correct?

**Comment by Ultrahead** [comment]
Is there any progress on this?

**Comment by joncampbell123** [comment]
> Is there any progress on this? [@joncampbell123](https://github.com/joncampbell123)

No, but I have been cleaning up the source code to make it more maintainable which I hope might make it easier to make a module from.

**Comment by Ultrahead** [comment]
DosBox (for Libreto) when working in Fullscreen does extend the viewport vertically adding two bars at the sides using surface mode in some Linux-based platforms. However, DosBox-x fails to behave the same way. I know scaling is not supported in surface mode, but it would be interesting to find out why and handy to fix it, when other output modes aren't an option. Note: using the SDL2 version.

**Comment by Ultrahead** [comment]
DosBox

![Image](https://github.com/user-attachments/assets/b2a428ac-f8fb-4165-b4a7-bfe585147151)

DosBox-x

![Image](https://github.com/user-attachments/assets/b5071812-708c-4939-a0c4-d0f201d9995d)
